File: waterContainerMaxArea.js

function maxArea(heights) {
    let maxArea = 0;
    let l = 0, r = heights.length - 1;

    while (l < r) {
        let area = Math.min(heights[r], heights[l]) * (r - l);
        maxArea = Math.max(maxArea, area);
        if (heights[r] > heights [l]) l += 1
        else r -= 1
    }
    return maxArea;
}

const heights = [1,8,6,2,5,4,8,3,7];
console.log(maxArea(heights));

File: waterContainerMaxArea.java

public class waterContainerMaxArea {
    public static int maxArea(int[] heights) {
        int res = 0;
        int l = 0;
        int r = heights.length - 1;

        while (l < r) {
            int area = Math.min(heights[l], heights[r]) * (r - l);
            res = Math.max(res, area);

            if (heights[l] < heights[r]) l++;
            else r--;
        }
        return res;
    }

    public static void main(String[] args) {
        int[] heights = {1, 8, 6, 2, 5, 4, 8, 3, 7};
        System.out.println(maxArea(heights));
    }


}